README.md
# Semantic Version
[![Go Report Card](https://goreportcard.com/badge/github.com/philhanna/semanticversion)][idGoReportCard]
[![PkgGoDev](https://pkg.go.dev/badge/github.com/philhanna/semanticversion)][idPkgGoDev]

Go implementation of a semantic versioned `Version` class.
See [https://semver.org][idSemver] for a normative description of semantic versioning.

## Overview
The basic structure of a semantic version is `major.minor.patch`,
where each component is an integer with no leading zeros.
These three components are all required.

A semantic version may also contain:
- A prerelease component, which follows the first '-' and ends
at the first '+', if any
- A buildmetadata component, which follows the first '+'
Both of these are optional.

## Semantics of each component

### Major
A change in the major version means that backwards incompatible
changes have been introduced into the public API.  When the
major version changes, the minor and patch versions must be
reset to zero.

### Minor
A change in the minor version means that backwards compatible
functionality has been added to the public API.  When the minor
version changes, the patch version must be reset to zero.

### Patch
A change in the patch version means that backwards compatible
bug fixes have been introduced.

See the [official documentation][idSemver] for a description of the
prerelease and buildmetadata components.

## Ordering
Semantic versions have a strict ordering; that is, any two
are comparable according to a set of rules:

1. Precedence MUST be calculated by separating the version into
major, minor, patch and pre-release identifiers in that order (Build
metadata does not figure into precedence).

2. Precedence is determined by the first difference when comparing
each of these identifiers from left to right as follows: Major,
minor, and patch versions are always compared numerically.

Example: 1.0.0 < 2.0.0 < 2.1.0 < 2.1.1.

3. When major, minor, and patch are equal, a pre-release version has
lower precedence than a normal version:

Example: 1.0.0-alpha < 1.0.0.

4. Precedence for two pre-release versions with the same major,
minor, and patch version MUST be determined by comparing each dot
separated identifier from left to right until a difference is found
as follows:

  - Identifiers consisting of only digits are compared numerically.

  - Identifiers with letters or hyphens are compared lexically in
    ASCII sort order.

  - Numeric identifiers always have lower precedence than
    non-numeric identifiers.

  - A larger set of pre-release fields has a higher precedence
    than a smaller set, if all of the preceding identifiers are
    equal.

    Example: 1.0.0-alpha < 1.0.0-alpha.1 < 1.0.0-alpha.beta
    < 1.0.0-beta < 1.0.0-beta.2 < 1.0.0-beta.11
    < 1.0.0-rc.1 < 1.0.0.

## Regular expression
See [https://regex101.com/r/Ly7O1x/3][idRegex]
for the regular expression used to parse the version string.

### Note
The website includes "99999999999999999999999.999999999999999999.99999999999999999"
as a valid Version string, but this contains components that overflow integer
literals in Go, even uint64. I have omitted this test case, but it would
seem to be irrelevant in actual usage.

[idSemver]: https://semver.org
[idRegex]: https://regex101.com/r/Ly7O1x/3
[idGoReportCard]: https://goreportcard.com/report/github.com/philhanna/semanticversion
[idPkgGoDev]: https://pkg.go.dev/github.com/philhanna/semanticversion

CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning].
The format is based on [Keep a Changelog].

## [Unreleased]
Added badges to `README.md`

## [v1.0.0] - 2023-02-19

This is a Go version of the original Python implementation,
which can still be found in the `python` branch.

[Semantic Versioning]: http://semver.org
[Keep a Changelog]: http://keepachangelog.com
[Unreleased]: https://github.com/philhanna/semanticVersion/compare/v1.0.0..HEAD 
[v1.0.0]: https://github.com/philhanna/semanticversion/compare/a14727b..v1.0.0


version_test.go
package semver

import (
	"testing"
)

func TestNewVersion(t *testing.T) {
	tests := []struct {
		versionString string
		want          Version
		isGood        bool
	}{
		{"bad", Version{}, false},
		{"1.0.0-alpha", Version{
			Major:         1,
			Minor:         0,
			Patch:         0,
			Prerelease:    "alpha",
			Buildmetadata: "",
			VersionString: "1.0.0-alpha",
		}, true},
	}
	for _, tt := range tests {
		// For convenience, all tests are named after their version
		// string
		name := tt.versionString
		t.Run(name, func(t *testing.T) {
			have, err := NewVersion(tt.versionString)
			if err == nil && !tt.isGood {
				t.Errorf("%s should have failed. have=%s", name, have)
			}
			if err != nil && tt.isGood {
				t.Errorf("%s failed but shouldn't have. have=%s", name, have)
			}
			want := tt.want
			if have != want {
				t.Errorf("have=%s,want=%v\n", have, want)
			}
		})
	}
}


ordering_test.go
package semver

import (
	"sort"
	"testing"
)

// Test cases for precedence, as outlined in the official documentation
// at https://semver.org.  The test case names refer to the examples
// given in the document.  I cover all the examples and have added test
// cases of my own.

func TestOrder_11_2(t *testing.T) {
	a, _ := NewVersion("1.0.0")
	b, _ := NewVersion("2.0.0")
	c, _ := NewVersion("2.1.0")
	d, _ := NewVersion("2.1.1")
	assertLess(t, a, b)
	assertLess(t, b, c)
	assertLess(t, c, d)
	assertGreaterOrEqual(t, d, c)
}

func TestOrder_11_3(t *testing.T) {
	a, _ := NewVersion("1.0.0-alpha")
	b, _ := NewVersion("1.0.0")
	assertLess(t, a, b)
}

func TestOrder_11_4(t *testing.T) {
	a, _ := NewVersion("1.0.0-alpha")
	b, _ := NewVersion("1.0.0-alpha.1")
	c, _ := NewVersion("1.0.0-alpha.beta")
	d, _ := NewVersion("1.0.0-beta")
	e, _ := NewVersion("1.0.0-beta.2")
	f, _ := NewVersion("1.0.0-beta.11")
	g, _ := NewVersion("1.0.0-rc.1")
	h, _ := NewVersion("1.0.0")
	assertLess(t, a, b)
	assertLess(t, b, c)
	assertLess(t, c, d)
	assertLess(t, d, e)
	assertLess(t, e, f)
	assertLess(t, f, g)
	assertLess(t, g, h)
}

func TestOrder_equalFull(t *testing.T) {
	a, _ := NewVersion("1.2.3-RC1+build3562")
	b, _ := NewVersion("1.2.3-RC2+build3562")
	assertEqual(t, a, a)
	assertNotEqual(t, a, b)
}

func TestOrder_equalNoBuildMetadata(t *testing.T) {
	a, _ := NewVersion("1.2.3-RC1")
	b, _ := NewVersion("1.2.3-RC2")
	assertEqual(t, a, a)
	assertNotEqual(t, a, b)
}

func TestOrder_equalNoPrerelease(t *testing.T) {
	a, _ := NewVersion("1.2.3+build3562")
	b, _ := NewVersion("1.2.3+build3563")
	assertEqual(t, a, a)
	assertEqual(t, a, b)
}

func TestOrder_compareLongerPrerelease(t *testing.T) {
	a, _ := NewVersion("1.0.0-alpha")
	b, _ := NewVersion("1.0.0-alpha.rc1")
	assertLess(t, a, b)
}

func TestOrder_compareWithPrerelease(t *testing.T) {
	a, _ := NewVersion("1.2.3----R-S.12.9.1--.12+meta")
	b, _ := NewVersion("1.1.7")
	assertGreater(t, a, b)
}

func TestOrder_singleDigitVsDoubleDigit(t *testing.T) {
	a, _ := NewVersion("2.3.4-10")
	b, _ := NewVersion("2.3.4-2")
	assertGreater(t, a, b)
}

func TestOrder_sort(t *testing.T) {
	expectedList := []string{
		"0.0.4",
		"1.0.0-0A.is.legal",
		"1.0.0-alpha",
		"1.0.0-alpha+beta",
		"1.0.0-alpha.1",
		"1.0.0-alpha.0valid",
		"1.0.0-alpha.beta",
		"1.0.0-alpha.beta.1",
		"1.0.0-alpha-a.b-c-somethinglong+build.1-aef.1-its-okay",
		"1.0.0-alpha0.valid",
		"1.0.0-beta",
		"1.0.0-rc.1+build.1",
		"1.0.0",
		"1.0.0+0.build.1-rc.10000aaa-kk-0.1",
		"1.1.2-prerelease+meta",
		"1.1.2+meta",
		"1.1.2+meta-valid",
		"1.1.7",
		"1.2.3----R-S.12.9.1--.12+meta",
		"1.2.3----RC-SNAPSHOT.12.9.1--.12+788",
		"1.2.3----RC-SNAPSHOT.12.9.1--.12",
		"1.2.3-SNAPSHOT-123",
		"1.2.3-beta",
		"1.2.3",
		"2.0.0-rc.1+build.123",
		"2.0.0",
		"2.0.0+build.1848",
		"2.0.1-alpha.1227",
		"10.2.3-DEV-SNAPSHOT",
		"10.20.30",
		"99999999999999999999999.999999999999999999.99999999999999999",
	}

	// Convert all version strings to Version types
	actualVersionList := make([]Version, 0, len(expectedList))
	for _, s := range expectedList {
		version, _ := NewVersion(s)
		actualVersionList = append(actualVersionList, version)
	}

	// Sort by the Version Compare function
	sort.Slice(actualVersionList, func(i, j int) bool {
		return actualVersionList[i].Compare(actualVersionList[j]) < 0
	})

	// Copy the versions back to a list of strings
	actualList := make([]string, 0, len(actualVersionList))
	for _, v := range actualVersionList {
		actualList = append(actualList, v.VersionString)
	}

	for i := range actualList {
		have := actualList[i]
		want := expectedList[i]
		if have != want {
			t.Errorf("At %d, have=%q, want=%q", i, have, want)
		}
	}
}

func assertLess(t *testing.T, a Version, b Version) {
	if !(a.Compare(b) < 0) {
		t.Errorf("%s is not less than %s", a, b)
	}
}

func assertGreater(t *testing.T, a Version, b Version) {
	if !(a.Compare(b) > 0) {
		t.Errorf("%s is not greater than %s", a, b)
	}
}

func assertEqual(t *testing.T, a Version, b Version) {
	if !(a.Compare(b) == 0) {
		t.Errorf("%s is not equal to %s", a, b)
	}
}

func assertGreaterOrEqual(t *testing.T, a Version, b Version) {
	if !(a.Compare(b) >= 0) {
		t.Errorf("%s is not greater than or equal to %s", a, b)
	}
}

func assertLessOrEqual(t *testing.T, a Version, b Version) {
	if !(a.Compare(b) <= 0) {
		t.Errorf("%s is not less than or equal to %s", a, b)
	}
}

func assertNotEqual(t *testing.T, a Version, b Version) {
	if !(a.Compare(b) != 0) {
		t.Errorf("%s is equal to %s", a, b)
	}
}


semantic_test.go
package semver

// Test data taken from https://regex101.com/r/Ly7O1x/3/

import "testing"

// TestVersion_good tests for valid version strings
func TestVersion_good(t *testing.T) {
	tests := []struct {
		versionString string
		major         int
		minor         int
		patch         int
		prerelease    string
		buildmetadata string
	}{
		{"0.0.4", 0, 0, 4, "", ""},
		{"1.2.3", 1, 2, 3, "", ""},
		{"10.20.30", 10, 20, 30, "", ""},
		{"1.1.2-prerelease+meta", 1, 1, 2, "prerelease", "meta"},
		{"1.1.2+meta", 1, 1, 2, "", "meta"},
		{"1.1.2+meta-valid", 1, 1, 2, "", "meta-valid"},
		{"1.0.0-alpha", 1, 0, 0, "alpha", ""},
		{"1.0.0-beta", 1, 0, 0, "beta", ""},
		{"1.0.0-alpha.beta", 1, 0, 0, "alpha.beta", ""},
		{"1.0.0-alpha.beta.1", 1, 0, 0, "alpha.beta.1", ""},
		{"1.0.0-alpha.1", 1, 0, 0, "alpha.1", ""},
		{"1.0.0-alpha0.valid", 1, 0, 0, "alpha0.valid", ""},
		{"1.0.0-alpha.0valid", 1, 0, 0, "alpha.0valid", ""},
		{"1.0.0-alpha-a.b-c-somethinglong+build.1-aef.1-its-okay", 1, 0, 0, "alpha-a.b-c-somethinglong",
			"build.1-aef.1-its-okay"},
		{"1.0.0-rc.1+build.1", 1, 0, 0, "rc.1", "build.1"},
		{"2.0.0-rc.1+build.123", 2, 0, 0, "rc.1", "build.123"},
		{"1.2.3-beta", 1, 2, 3, "beta", ""},
		{"10.2.3-DEV-SNAPSHOT", 10, 2, 3, "DEV-SNAPSHOT", ""},
		{"1.2.3-SNAPSHOT-123", 1, 2, 3, "SNAPSHOT-123", ""},
		{"1.0.0", 1, 0, 0, "", ""},
		{"2.0.0", 2, 0, 0, "", ""},
		{"1.1.7", 1, 1, 7, "", ""},
		{"2.0.0+build.1848", 2, 0, 0, "", "build.1848"},
		{"2.0.1-alpha.1227", 2, 0, 1, "alpha.1227", ""},
		{"1.0.0-alpha+beta", 1, 0, 0, "alpha", "beta"},
		{"1.2.3----RC-SNAPSHOT.12.9.1--.12+788", 1, 2, 3, "---RC-SNAPSHOT.12.9.1--.12", "788"},
		{"1.2.3----R-S.12.9.1--.12+meta", 1, 2, 3, "---R-S.12.9.1--.12", "meta"},
		{"1.2.3----RC-SNAPSHOT.12.9.1--.12", 1, 2, 3, "---RC-SNAPSHOT.12.9.1--.12", ""},
		{"1.0.0+0.build.1-rc.10000aaa-kk-0.1", 1, 0, 0, "", "0.build.1-rc.10000aaa-kk-0.1"},
		{"1.0.0-0A.is.legal", 1, 0, 0, "0A.is.legal", ""},
	}
	for _, tt := range tests {
		t.Run(tt.versionString, func(t *testing.T) {
			have, _ := NewVersion(tt.versionString)
			if have.Major != tt.major {
				t.Errorf("Major: have=%d,want=%d", have.Major, tt.major)
			}
			if have.Minor != tt.minor {
				t.Errorf("Minor: have=%d,want=%d", have.Minor, tt.minor)
			}
			if have.Patch != tt.patch {
				t.Errorf("Patch: have=%d,want=%d", have.Patch, tt.patch)
			}
			if have.Prerelease != tt.prerelease {
				t.Errorf("Prerelease: have=%q,want=%q", have.Prerelease, tt.prerelease)
			}
			if have.Buildmetadata != tt.buildmetadata {
				t.Errorf("Buildmetadata: have=%q,want=%q", have.Buildmetadata, tt.buildmetadata)
			}
		})
	}
}

// TestVersion_bad tests for invalid version strings
func TestVersion_bad(t *testing.T) {
	testCases := []string{
		"2.1",
		"1",
		"1.2",
		"1.2.3-0123",
		"1.2.3-0123.0123",
		"1.1.2+.123",
		"+invalid",
		"-invalid",
		"-invalid+invalid",
		"-invalid.01",
		"alpha",
		"alpha.beta",
		"alpha.beta.1",
		"alpha.1",
		"alpha+beta",
		"alpha_beta",
		"alpha.",
		"alpha..",
		"beta",
		"1.0.0-alpha_beta",
		"-alpha.",
		"1.0.0-alpha..",
		"1.0.0-alpha..1",
		"1.0.0-alpha...1",
		"1.0.0-alpha....1",
		"1.0.0-alpha.....1",
		"1.0.0-alpha......1",
		"1.0.0-alpha.......1",
		"01.1.1",
		"1.01.1",
		"1.1.01",
		"1.2",
		"1.2.3.DEV",
		"1.2-SNAPSHOT",
		"1.2.31.2.3----RC-SNAPSHOT.12.09.1--..12+788",
		"1.2-RC-SNAPSHOT",
		"-1.0.3-gamma+b7718",
		"+justmeta",
		"9.8.7+meta+meta",
		"9.8.7-whatever+meta+meta",
		"99999999999999999999999.999999999999999999.99999999999999999----RC-SNAPSHOT.12.09.1",
		"--------------------------------..12",
	}
	for _, versionString := range testCases {
		t.Run(versionString, func(t *testing.T) {
			_, err := NewVersion(versionString)
			if err == nil {
				t.Errorf("%q should have been an invalid version string", versionString)
			}
		})
	}
}


version.go
// A semantic versioned version class
//
// See https://semver.org for a normative description of
// semantic versioning.
//
// The basic structure of a semantic version is major.minor.patch,
// where each component is an integer with no leading zeros.
// These three components are all required.
//
// A semantic version may also contain:
// - A prerelase component, which follows the first '-' and ends
// at the first '+', if any
// - A buildmetadata component, which follows the first '+'
// Both of these are optional.
//
// A change in the major version means that backwards incompatible
// changes have been introduced into the public API.  When the
// major version changes, the minor and patch versions must be
// reset to zero.
//
// A change in the minor version means that backwards compatible
// functionality has been added to the public API.  When the minor
// version changes, the patch version must be reset to zero.
//
// A change in the patch version means that backwards compatible
// bug fixes have been introduced.
//
// See the official documentation for a description of the
// prerelease and buildmetadata components.
//
// Semantic versions have a strict ordering; that is, any two
// are comparable according to a set of rules:
//
// 1. Precedence MUST be calculated by separating the version into
// major, minor, patch and pre-release identifiers in that order (Build
// metadata does not figure into precedence).
//
// 2. Precedence is determined by the first difference when comparing
// each of these identifiers from left to right as follows: Major,
// minor, and patch versions are always compared numerically.
//
// Example: 1.0.0 < 2.0.0 < 2.1.0 < 2.1.1.
//
// 3. When major, minor, and patch are equal, a pre-release version has
// lower precedence than a normal version:
//
// Example: 1.0.0-alpha < 1.0.0.
//
// 4. Precedence for two pre-release versions with the same major,
// minor, and patch version MUST be determined by comparing each dot
// separated identifier from left to right until a difference is found
// as follows:
//
//   - Identifiers consisting of only digits are compared numerically.
//
//   - Identifiers with letters or hyphens are compared lexically in
//     ASCII sort order.
//
//   - Numeric identifiers always have lower precedence than
//     non-numeric identifiers.
//
//   - A larger set of pre-release fields has a higher precedence
//     than a smaller set, if all of the preceding identifiers are
//     equal.
//
//     Example: 1.0.0-alpha < 1.0.0-alpha.1 < 1.0.0-alpha.beta
//     < 1.0.0-beta < 1.0.0-beta.2 < 1.0.0-beta.11
//     < 1.0.0-rc.1 < 1.0.0.
//
// See https://regex101.com/r/Ly7O1x/3/ for the regular expression used
// to parse the version string.
package semver

import (
	"errors"
	"fmt"
	"regexp"
	"strconv"
	"strings"
)

// ---------------------------------------------------------------------
// Type definitions
// ---------------------------------------------------------------------

// Version contains the parsed components of a version string.
type Version struct {
	Major         int
	Minor         int
	Patch         int
	Prerelease    string
	Buildmetadata string
	VersionString string
}

// ---------------------------------------------------------------------
// Constructors
// ---------------------------------------------------------------------

// NewVersion parses a version string and returns a new Version from it.
//
// NOTE: The regular expression defined at https://semver.org does not
// include a leading "v", but this constructor does, for convenience in
// working with Go best practices.
func NewVersion(vs string) (Version, error) {
	p := new(Version)
	p.VersionString = vs
	reParts := []string{
		`^`,
		`v?`,
		`(?P<major>0|[1-9]\d*)`,
		`\.`,
		`(?P<minor>0|[1-9]\d*)`,
		`\.`,
		`(?P<patch>0|[1-9]\d*)`,
		`(?:-(?P<prerelease>(?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*)(?:\.(?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*))*))?(?:\+`,
		`(?P<buildmetadata>[0-9a-zA-Z-]+(?:\.[0-9a-zA-Z-]+)*))?$`,
	}
	reString := strings.Join(reParts, "")
	re := regexp.MustCompile(reString)
	tokens := re.FindStringSubmatch(vs)
	if tokens == nil {
		errmsg := fmt.Sprintf("Invalid version string: %s", vs)
		return Version{}, errors.New(errmsg)
	}

	// Populate the structure
	p.Major, _ = strconv.Atoi(tokens[1])
	p.Minor, _ = strconv.Atoi(tokens[2])
	p.Patch, _ = strconv.Atoi(tokens[3])
	p.Prerelease = tokens[4]
	p.Buildmetadata = tokens[5]

	return *p, nil
}

// ---------------------------------------------------------------------
// Methods
// ---------------------------------------------------------------------

// Compare compares two versions and returns -1, 0, or 1, depending on
// whether this version is less than, equal to, or greater than the
// other version. The comparisons are done according to Section 11 of
// the specification.
//
// Note that build metadata is NOT considered in precedence comparisons,
// according to the specification.
func (self Version) Compare(other Version) int {
	switch {
	case self.Major < other.Major:
		return -1
	case self.Major > other.Major:
		return 1
	case self.Minor < other.Minor:
		return -1
	case self.Minor > other.Minor:
		return 1
	case self.Patch < other.Patch:
		return -1
	case self.Patch > other.Patch:
		return 1
	}

	// At this point, we know the three primary components are equal.

	// 11.3 A pre-release version has lower precedence than a normal
	// version

	if self.Prerelease == "" && other.Prerelease == "" {
		return 0
	}
	if self.Prerelease != "" && other.Prerelease == "" {
		return -1
	}
	if self.Prerelease == "" && other.Prerelease != "" {
		return 1
	}

	// 11.4 - Precedence for two pre-release versions with the same
	// major, minor, and patch must be determined by comparing each
	// dot-separated identifier from left to right until a difference is
	// found

	theseFields := strings.Split(self.Prerelease, ".")
	thoseFields := strings.Split(other.Prerelease, ".")

	// Need to make these two lists the same length for joint iteration
	for len(theseFields) < len(thoseFields) {
		theseFields = append(theseFields, "")
	}
	for len(theseFields) > len(thoseFields) {
		thoseFields = append(thoseFields, "")
	}

	for i := range theseFields {
		this := theseFields[i]
		that := thoseFields[i]

		// 11.4.4 Longer sets are greater than shorter sets
		if this == "" && that != "" {
			return -1
		}
		if this != "" && that == "" {
			return 1
		}

		if this != "" && that != "" {
			// 11.4.1 Integers are compared numerically
			if isNumeric(this) && isNumeric(that) {
				intthis, _ := strconv.Atoi(this)
				intthat, _ := strconv.Atoi(that)
				if intthis < intthat {
					return -1
				}
				if intthis > intthat {
					return 1
				}
			}
			// 11.4.3 Numerics are always less than non-numerics
			if isNumeric(this) && !isNumeric(that) {
				return -1
			}
			if !isNumeric(this) && isNumeric(that) {
				return 1
			}
			// 11.4.2 Nonnumeric fields are compared lexicographically
			if this < that {
				return -1
			}
			if this > that {
				return 1
			}
		}
	}

	// Welp, they must be equal
	return 0
}

// String returns a string representation of this Version
func (v Version) String() string {
	parts := []string{
		fmt.Sprintf("Major: %d", v.Major),
		fmt.Sprintf("Minor: %d", v.Minor),
		fmt.Sprintf("Patch: %d", v.Patch),
		fmt.Sprintf("Prerelease: %q", v.Prerelease),
		fmt.Sprintf("Buildmetadata: %q", v.Buildmetadata),
	}
	s := "Version{" + strings.Join(parts, ", ") + "}"
	return s
}

// ---------------------------------------------------------------------
// Functions
// ---------------------------------------------------------------------

// isNumeric returns true if the string is composed entirely of digits
func isNumeric(s string) bool {
	for _, ch := range s {
		if ch >= '0' && ch <= '9' {
			continue
		}
		return false
	}
	return true
}


semver/version.py
import re
from functools import total_ordering
from itertools import zip_longest


def is_numeric(arg: str) -> bool:
    return all([c.isdigit() for c in arg])


@total_ordering
class Version:
    """ A semantic versioned version class

    See https://semver.org for a normative description of
    semantic versioning.

    The basic structure of a semantic version is major.minor.patch,
    where each component is an integer with no leading zeros.
    These three components are all required.

    A semantic version may also contain:
    - A prerelase component, which follows the first '-' and ends
    at the first '+', if any
    - A buildmetadata component, which follows the first '+'
    Both of these are optional.

    A change in the major version means that backwards incompatible
    changes have been introduced into the public API.  When the
    major version changes, the minor and patch versions must be
    reset to zero.

    A change in the minor version means that backwards compatible
    functionality has been added to the public API.  When the minor
    version changes, the patch version must be reset to zero.

    A change in the patch version means that backwards compatible
    bug fixes have been introduced.

    See the official documentation for a description of the
    prerelease and buildmetadata components.

    Semantic versions have a strict ordering; that is, any two
    are comparable according to a set of rules:

    1. Precedence MUST be calculated by separating the version into
    major, minor, patch and pre-release identifiers in that order (Build
    metadata does not figure into precedence).

    2. Precedence is determined by the first difference when comparing
    each of these identifiers from left to right as follows: Major,
    minor, and patch versions are always compared numerically.

    Example: 1.0.0 < 2.0.0 < 2.1.0 < 2.1.1.

    3. When major, minor, and patch are equal, a pre-release version has
    lower precedence than a normal version:

    Example: 1.0.0-alpha < 1.0.0.

    4. Precedence for two pre-release versions with the same major,
    minor, and patch version MUST be determined by comparing each dot
    separated identifier from left to right until a difference is found
    as follows:

        - Identifiers consisting of only digits are compared numerically.

        - Identifiers with letters or hyphens are compared lexically in
          ASCII sort order.

        - Numeric identifiers always have lower precedence than
          non-numeric identifiers.

        - A larger set of pre-release fields has a higher precedence
          than a smaller set, if all of the preceding identifiers are
          equal.

          Example: 1.0.0-alpha < 1.0.0-alpha.1 < 1.0.0-alpha.beta
                  < 1.0.0-beta < 1.0.0-beta.2 < 1.0.0-beta.11
                  < 1.0.0-rc.1 < 1.0.0.


    See https://regex101.com/r/Ly7O1x/3/ for the regular expression used
    to parse the version string.

    """

    def __init__(self, version_string: str):
        self._version_string = version_string
        regexp = re.compile((
            r'''^'''
            r'''(?P<major>0|[1-9]\d*)'''
            r'''\.'''
            r'''(?P<minor>0|[1-9]\d*)'''
            r'''\.'''
            r'''(?P<patch>0|[1-9]\d*)'''
            r'''(?:-(?P<prerelease>(?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*)(?:\.(?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*))*))?'''
            r'''(?:\+(?P<buildmetadata>[0-9a-zA-Z-]+(?:\.[0-9a-zA-Z-]+)*))'''
            r'''?$'''
        ))
        m = regexp.match(version_string)
        if not m:
            errmsg = f"{version_string} is not a valid semantic version number. See https://semver.org"
            raise ValueError(errmsg)
        self._major: int = int(m.group('major'))
        self._minor: int = int(m.group('minor'))
        self._patch: int = int(m.group('patch'))
        self._prerelease = m.group('prerelease')
        self._buildmetadata = m.group('buildmetadata')

    @property
    def version_string(self):
        return self._version_string

    @property
    def major(self):
        """ Returns the major version """
        return self._major

    @property
    def minor(self):
        """ Returns the minor version """
        return self._minor

    @property
    def patch(self):
        """ Returns the patch version """
        return self._patch

    @property
    def prerelease(self):
        """ Returns the prerelease version """
        return self._prerelease

    @property
    def buildmetadata(self):
        """ Returns the buildmetadata version """
        return self._buildmetadata

    def __str__(self) -> str:
        basic_string = ".".join([str(self.major), str(self.minor), str(self.patch)])
        if self.prerelease is not None:
            basic_string += "-" + self.prerelease
        if self.buildmetadata is not None:
            basic_string += "+" + self.buildmetadata
        return basic_string

    def __gt__(self, other):
        return self.compare(other) > 0

    def __eq__(self, other):
        return self.compare(other) == 0

    def __hash__(self):
        return hash(self.__str__())

    def compare(self, other):
        """ Compares two versions and returns -1, 0, or 1

        The __gt__, __lt__, __eq__, __ge__, __le__, and __ne__
        relations are based on this.

        Note that build metadata is NOT considered in precedence comparisons,
        according to the specification.
        """
        if self.major < other.major:
            return -1
        elif self.major > other.major:
            return 1
        elif self.minor < other.minor:
            return -1
        elif self.minor > other.minor:
            return 1
        elif self.patch < other.patch:
            return -1
        elif self.patch > other.patch:
            return 1

        # At this point, we know the three primary components are equal

        # 11.3 A pre-release version has lower precedence than a normal version
        if all([self.prerelease is not None,
                other.prerelease is None]):
            return -1
        if all([self.prerelease is None,
                other.prerelease is not None]):
            return 1
        if all([self.prerelease is None,
                other.prerelease is None]):
            return 0

        # 11.4 - Precedence for two pre-release versions with the same
        # major, minor, and patch must be determined by comparing each
        # dot-separated identifier from left to right until a difference
        # is found

        these_fields = self.prerelease.split('.')
        those_fields = other.prerelease.split('.')
        for this, that in zip_longest(these_fields, those_fields):

            # 11.4.4 Longer sets are greater than shorter sets
            if this is None and that is not None:
                return -1
            if this is not None and that is None:
                return 1

            # 11.4.1 Integers are compared numerically
            if this and that:
                if is_numeric(this) and is_numeric(that):
                    intthis = int(this)
                    intthat = int(that)
                    if intthis < intthat:
                        return -1
                    if intthis > intthat:
                        return 1

            # 11.4.3 Numerics are always less than non-numerics
            if this and that:
                if is_numeric(this) and not is_numeric(that):
                    return -1
                if not is_numeric(this) and is_numeric(that):
                    return 1

            # 11.4.2 Nonnumeric fields are compared lexicographically
            if this < that:
                return -1
            if this > that:
                return 1

        # Welp, they must be equal

        return 0
